cmake/Modules/FindMKL.cmake

```diff
@@ -1,120 +1,205 @@
-  endif()
-
-  find_library(${__mkl_lib_upper}_LIBRARY
-        NAMES ${__mkl_lib}
-        PATHS ${MKL_ROOT} "${MKL_INCLUDE_DIR}/.."
-        PATH_SUFFIXES ${__path_suffixes}
-        DOC "The path to Intel(R) MKL ${__mkl_lib} library")
-  mark_as_advanced(${__mkl_lib_upper}_LIBRARY)
-
-  list(APPEND __looked_for ${__mkl_lib_upper}_LIBRARY)
-  list(APPEND MKL_LIBRARIES ${${__mkl_lib_upper}_LIBRARY})
-endforeach()
-
-
-if(NOT MKL_USE_SINGLE_DYNAMIC_LIBRARY)
-  if (MKL_USE_STATIC_LIBS)
-    set(__iomp5_libs iomp5 libiomp5mt.lib)
-  else()
-    set(__iomp5_libs iomp5 libiomp5md.lib)
-  endif()
-
-  if(WIN32)
-    find_path(INTEL_INCLUDE_DIR omp.h PATHS ${INTEL_ROOT} PATH_SUFFIXES include)
-    list(APPEND __looked_for INTEL_INCLUDE_DIR)
-  endif()
-
-  find_library(MKL_RTL_LIBRARY ${__iomp5_libs}
-     PATHS ${INTEL_RTL_ROOT} ${INTEL_ROOT}/compiler ${MKL_ROOT}/.. ${MKL_ROOT}/../compiler
-     PATH_SUFFIXES ${__path_suffixes}
-     DOC "Path to OpenMP runtime library")
-
-  list(APPEND __looked_for MKL_RTL_LIBRARY)
-  list(APPEND MKL_LIBRARIES ${MKL_RTL_LIBRARY})
-endif()
-
-
-include(FindPackageHandleStandardArgs)
-find_package_handle_standard_args(MKL DEFAULT_MSG ${__looked_for})
-
-if(MKL_FOUND)
-  message(STATUS "Found MKL (include: ${MKL_INCLUDE_DIR}, lib: ${MKL_LIBRARIES}")
-endif()
-
-caffe_clear_vars(__looked_for __mkl_libs __path_suffixes __lib_suffix __iomp5_libs)
+# Find the MKL libraries
+#
+# Options:
+#
+#   MKL_USE_SINGLE_DYNAMIC_LIBRARY  : use single dynamic library interface
+#   MKL_USE_STATIC_LIBS             : use static libraries
+#   MKL_MULTI_THREADED              : use multi-threading
+#   MKL_USE_IDEEP                   : use IDEEP interface
+#   MKL_USE_MKLML                   : use MKLML interface
+#
+# This module defines the following variables:
+#
+#   MKL_FOUND            : True mkl is found
+#   MKL_INCLUDE_DIR      : unclude directory
+#   MKL_LIBRARIES        : the libraries to link against.
+
+# ---[ Options
+include(CMakeDependentOption)
+option(MKL_USE_IDEEP "Use IDEEP interface" ON)
+option(MKL_USE_MKLML "Use MKLML interface" ON)
+
+if(MKL_USE_IDEEP)
+  set(IDEEP_ROOT "${PROJECT_SOURCE_DIR}/third_party/ideep")
+  set(MKLDNN_ROOT "${IDEEP_ROOT}/mkl-dnn")
+  set(__ideep_looked_for IDEEP_ROOT)
+
+  find_path(IDEEP_INCLUDE_DIR ideep.hpp PATHS ${IDEEP_ROOT} PATH_SUFFIXES include)
+  find_path(MKLDNN_INCLUDE_DIR mkldnn.hpp mkldnn.h PATHS ${MKLDNN_ROOT} PATH_SUFFIXES include)
+  if (NOT MKLDNN_INCLUDE_DIR)
+    execute_process(COMMAND git submodule update --init mkl-dnn WORKING_DIRECTORY ${IDEEP_ROOT})
+    find_path(MKLDNN_INCLUDE_DIR mkldnn.hpp mkldnn.h PATHS ${MKLDNN_ROOT} PATH_SUFFIXES include)
+  endif()
+
+  if (MKLDNN_INCLUDE_DIR)
+    # to avoid adding conflicting submodels
+    set(ORIG_WITH_TEST ${WITH_TEST})
+    set(WITH_TEST OFF)
+    add_subdirectory(${IDEEP_ROOT})
+    set(WITH_TEST ${ORIG_WITH_TEST})
+
+    file(GLOB_RECURSE MKLML_INCLUDE_DIR ${MKLDNN_ROOT}/external/*/mkl_vsl.h)
```

Comment:
Why would we have `MKLML_INCLUDE_DIR` when processing `MKLDNN_INCLUDE_DIR`?

Reply:
MKLDNN depends on part of MKLML

Reply:
:O

Reply:
The dependence is because the GEMM part of MKL-ML is still a bit faster than the open source implementation in MKL-DNN. We expect this dependence to be removed in the future after the GEMM performance in MKL-DNN is on par with MKL-ML.